File: leetcode/C++/Next_permutation.cpp
class Solution {
	public:
		void nextPermutation(vector<int>& nums) {
			if(!nums.size()) return ;
			cal_next(nums, nums.size()-1);

		}

		void cal_next(vector<int> & nums, int end)
		{
			if(end < 0)
			{
				sort(nums.begin(), nums.end());
				return ;
			}

			for(int i = nums.size()-1; i >= end; i--)
			{
				if(nums[i] > nums[end])
				{
					swap(nums[i], nums[end]);
					sort(nums.begin()+end+1, nums.end());
					return ;
				}
			}
			cal_next(nums, end-1);
		}
};

File: leetcode/array/ Next Permutation.go
func nextPermutation(nums []int)  {
    lowestj, lowesti := -1, -1
    for i := len(nums)-1; i >= 0; i--{
        for j := i; j >= 0; j--{
            if nums[j] < nums[i] && j > lowestj{
                lowestj = j 
                lowesti = i
            }
        }
    }
    if lowestj != -1{
        nums[lowesti], nums[lowestj] = nums[lowestj], nums[lowesti]
        sort.Ints(nums[lowestj+1:])
        return 
    }
    reverse(nums)
    return 
}